move sessions to connection_manager

czue opened the pull request:
@snopoke looking for feedback on this

might be easier to review commit by commit

czue committed:
move connection manager to corehq.db

czue committed:
move existing sessions to connection_manager

czue committed:
move test utilities to utils module

czue committed:
fix tests

removes unneeded dispose_all calls and adds Session.remove calls that
are needed

dimagimon commented on corehq/apps/userreports/tasks.py:
```diff
@@ -5,7 +5,8 @@
 from corehq.apps.userreports.models import DataSourceConfiguration, CustomDataSourceConfiguration
-from corehq.apps.userreports.sql import IndicatorSqlAdapter, create_engine
+from corehq.apps.userreports.sql import IndicatorSqlAdapter
+from corehq.db import connection_manager, create_engine
```
F401 'connection_manager' imported but unused

czue commented on corehq/apps/userreports/tests/test_columns.py:
```diff
@@ -78,10 +77,6 @@ def testBadFormat(self):
 
-    @classmethod
-    def tearDownClass(cls):
-        connection_manager.dispose_all()
```
now handled by testrunner

czue commented on corehq/apps/userreports/sql/connection.py:
```diff
@@ -1,54 +1,4 @@
-import sqlalchemy
```
most of this just moved - i'll comment on interesting bits

czue commented on corehq/apps/userreports/tests/test_data_source_config.py:
```diff
@@ -71,45 +68,6 @@ def test_indicators(self, datetime_mock):
                 self.assertEqual(str(expected_indicators[result.column.id]), result.value)
 
 
-def get_sample_data_source():
```
just moved with no changes

czue commented on corehq/apps/userreports/tests/test_view.py:
```diff
@@ -133,6 +133,9 @@ def _delete_everything(cls):
         cls._delete_everything()
+        # todo: understand why this is necessary. the view call uses the session and the
+        # signal doesn't fire to kill it.
+        Session.remove()
```
@snopoke any thoughts on this? I had to add it to avoid having the "still open connections" issue after tests finished. i think the scope of the test view is different from the actual test cases somehow, and the signal doesn't fire. what's odd is that it used to work until this change.

snopoke replied:
Is it not because you've removed `connection_manager.dispose_all()` from other test tear down methods?  When you say the signal doesn't fire, do you mean the  `request_finished` signal? Perhaps its this test isn't using the request cycle to test the view?

I don't thing `connection_manager.dispose_all()` should be required in tearDown methods but it might be necessary to have `Session.remove()`

snopoke replied:
Just thinking about this again, do all tests get run in the same thread?

czue replied:
the `testrunner` disposing and session closing seems to work for everything except this. i was wondering if the test client somehow runs in a separate thread though i don't see any evidence of that [in the docs](https://docs.djangoproject.com/en/1.8/topics/testing/tools/#the-test-client)

czue commented on corehq/db.py:
```diff
@@ -1,13 +1,82 @@
+    return sqlalchemy.create_engine(connection_string)
+
+
+class SessionFactory(object):
```
review

czue commented on corehq/db.py:
```diff
@@ -1,13 +1,82 @@
+            self._session_factories[engine_id] = SessionFactory(self._get_connection_string(engine_id))
+        return self._session_factories[engine_id]
+
+    def get_session_constructor(self, engine_id=DEFAULT_ENGINE_ID):
```
wasn't sure what to call this exactly. returns what used to be `db.Session`

snopoke replied:
maybe `get_session_factory`?

czue replied:
went with `get_scoped_session` since i was calling the other thing the session factory. happy to change it if you want though (but then i want to come up with another name for the class)

czue commented on testrunner.py:
```diff
@@ -206,9 +206,10 @@ def run_db_tests(self, suite):
 
-        from corehq.db import _engine, Session
+        from corehq.db import Session, connection_manager
         Session.remove()
```
this seems to do the trick except in a test view call that uses a session

snopoke replied:
Maybe interaction with other tests?

czue commented on corehq/db.py:
```diff
@@ -1,13 +1,82 @@
+
+
+connection_manager = ConnectionManager()
+Session = connection_manager.get_session_constructor(DEFAULT_ENGINE_ID)
```
for backwards compatibility and convenience

czue committed:
[ci skip] unused import

czue commented on corehq/db.py:
```diff
@@ -1,13 +1,82 @@
+        If found, disposes an engine and removes it from this, otherwise does nothing.
+        """
+        if engine_id in self._session_factories:
+            factory = self._session_factories.pop(engine_id)
```
it wasn't clear to me whether this should also call `factory.Session.remove()`. thoughts?

snopoke replied:
Yes, we should definitely call Session.remove() first. Extract form the docs:

"Connections that are checked out are not discarded when the engine is disposed or garbage collected, as these connections are still strongly referenced elsewhere by the application. However, after Engine.dispose() is called, those connections are no longer associated with that Engine; when they are closed, they will be returned to their now-orphaned connection pool which will ultimately be garbage collected, once all connections which refer to it are also no longer referenced anywhere. Since this process is not easy to control, it is strongly recommended that Engine.dispose() is called only after all checked out connections are checked in or otherwise de-associated from their pool." 
http://docs.sqlalchemy.org/en/rel_0_9/core/connections.html#engine-disposal

snopoke replied:
Note that `Session.remove()` will only release the connection for the current thread so its possible that there could be other threads that are still holding connections and I don't think there is a way to force release them.

czue replied:
yeah, i figured that we would mostly rely on other mechanisms to explicitly close them, but seems fine/safe to also call it here

snopoke commented:
Looks good to me

czue committed:
rename function, close sessions when disposing engine

czue commented:
ok, made some tweaks and i think this is good to go pending feedback/tests

czue committed:
remove unnecessary dispose call

czue committed:
don't use a new engine in the rebuild task

czue committed:
remove no-longer-used engine param

czue committed:
use table-specific sessions for view queries

czue committed:
close all sessions at the end of each view

snopoke commented on corehq/db.py:
```diff
@@ -1,13 +1,92 @@
+    def get_scoped_session(self, engine_id=DEFAULT_ENGINE_ID):
+        """
+        This returns a class that can be instantiated and will have a session scoped
+        to the local thread.
```
After reading this: http://docs.sqlalchemy.org/en/rel_1_0/orm/contextual.html#implicit-method-access I think you could just say that this returns a scoped session object.

snopoke commented:
Build stalled - kicked it :mans_shoe:

czue committed:
Merge pull request #7490 from dimagi/session-continuation

Session continuation

dimagimon commented on corehq/apps/userreports/tasks.py:
```diff
@@ -5,7 +5,8 @@
 from corehq.apps.userreports.models import DataSourceConfiguration, CustomDataSourceConfiguration
-from corehq.apps.userreports.sql import IndicatorSqlAdapter, create_engine
+from corehq.apps.userreports.sql import IndicatorSqlAdapter
+from corehq.db import create_engine
```
F401 'create_engine' imported but unused

dimagimon commented on corehq/db.py:
```diff
@@ -1,19 +1,104 @@
+        return self._get_or_create_factory(engine_id).engine
+
+
+    def close_scoped_sessions(self):
```
E303 too many blank lines (2)

czue commented:
(i'm going to respond to a few of your comments and change a few more things so no worries on merge for now)

czue committed:
renames and support different databases in sqlreport

- SessionFactory --> SessionHelper
- get_session_factory --> get_scoped_session

czue committed:
unused import

czue committed:
better name

czue commented:
ok, i think this is good to go when tests pass if you're good.

czue committed:
[ci skip] use engine_id

czue commented:
:white_check_mark: and updated (i decided to save travis some cycles on the [last commit](https://github.com/dimagi/commcare-hq/commit/5ffc86ebef66b8e914b0fc964efb79ea496a9032))

snopoke commented:
:+1:
